### ISSUE TITLE
[do not close] NGA/task 3: tracking sorting task

### ISSUE THREAD

**Issue by fnrizzi** [kokkos/kokkos#5071]
- [ ] 3.1 Make top level `Kokkos::sort(Kokkos::View)` map toThrust/RocThrust/std::sort/OneAPI as appropriate
   (@jstrzebonski)
- [ ] 3.2 Explore generic implementation of RadixSort and BitonicSort (@PhilMiller)
   - report on potential issues in API design due to need of temporary allocations etc.
   - implement at top level and compare to Thrust/RocThrust 
   - implementation working at Team and Potential Thread level in hierarchical parallelism
   - will potentially down-select optimal algorithm
- [ ] 3.3 (Stretch) Integrate into mainline Kokkos as appropriate with tests (assign to: TBD)

**Comment by PhilMiller** [comment]
Here's a rough list of open issues in Kokkos relating to sorting algorithms and their use cases, to inform 3.2
* #645 
* #1931
* #2914
* #3089
* #3285 
* #4487 
* #4978

**Comment by PhilMiller** [comment]
Possibly related issues in Kokkos Kernels:
* https://github.com/kokkos/kokkos-kernels/issues/10
* https://github.com/kokkos/kokkos-kernels/issues/473

There's various code in Kokkos and Kokkos Kernels that implement various sorting and sorting-adjacent functionality

**Comment by PhilMiller** [comment]
A thought inspired by the Python standard library's sorting function:

It can be quite valuable to distinguish "compute the comparison key of element `i`" from "compare (the keys of) elements `i` and `j`"

In at least some of our prospective use cases, this would potentially let us still use an integer sorting algorithm (like Radix Sort) on non-integer values. E.g. in the bin-particles-by-cell case, the notional key function returns an integer cell index

**Comment by PhilMiller** [comment]
A couple of earlier issues mention interfaces like `sort2` and `sort3` that sort several parallel arrays based on the ordering of one that provides the keys. Here in C++-land, for use cases that don't need to get back the permutation, there's no reason we can't have `sort(Comparator, KeyArray, ValueArrays...)` for an arbitrary number of parallel arrays to be sorted.

Hypothetically, this could even be implemented a bit more efficiently than permuting each array one by one, since we can re-use the loads of the permutation or even re-compute the sort of the key array, depending on pragmatics of the operation in question.

**Comment by PhilMiller** [comment]
Regarding temporary memory allocations, potential large items include a length-`N` permutation array of indices, an inversion of the permutation array, and a length-`N` buffer array of value type elements to do an out-of-place application of the permutation, and then swap the pointer into place.

**Comment by PhilMiller** [comment]
> A couple of earlier issues mention interfaces like `sort2` and `sort3` that sort several parallel arrays based on the ordering of one that provides the keys. Here in C++-land, for use cases that don't need to get back the permutation, there's no reason we can't have `sort(Comparator, KeyArray, ValueArrays...)` for an arbitrary number of parallel arrays to be sorted.
> 
> Hypothetically, this could even be implemented a bit more efficiently than permuting each array one by one, since we can re-use the loads of the permutation or even re-compute the sort of the key array, depending on pragmatics of the operation in question.

Or, crossing the streams a bit, if we have a temporary array to copy values in/out of to apply the permutation, and it's sized for N elements of the maximum size (e.g. 64 bits), N smaller elements that can pack into that (2x 32 bits) could occupy it concurrently

**Comment by PhilMiller** [comment]
A further thought on applicability of 'integer' (sub-comparison-bound) sorting algorithms like RadixSort - there's no reason we can't add support for composite keys (say in a `pair` or `tuple` or `std::get`able struct) each of whose elements is interpretable as raw bits. Again, if the input are some more complicated object, but a 'key function' can produce such a composite, then this becomes pretty broadly applicable. EMPIRE's full sorting would work well with this, for instance.

**Comment by fnrizzi** [comment]
@PhilMiller thanks for starting on this! 
I think these are good general thoughts. Following up on this, can you please elaborate more on "explore generic implementation of RadixSort and BitonicSort"? What I suggest would be to do some research and come up a summary that covers: 
- what opensource libraries offer these sorting algorithms 
- which implementations are most commonly used and what are the nuisances/details  
- typical use cases one would use radix vs bitonic for

**Comment by brian-kelley** [comment]
@fnrizzi @PhilMiller I just opened #5317 with team and thread level sort and sort_by_key (always using bitonic, so it's in-place). This impl is mostly from KokkosKernels, but with a nicer interface.

At one point we tried to publish an overview paper about all of KokkosKernels, and I ran performance experiments on that. For sorting 256-element arrays in shared memory, this was about 9% slower than CUB's block radix sort. I think (for most users at least), the flexibility of bitonic is a good tradeoff. CUB makes you allocate both shared and register/local space for the radix, and of course doesn't allow custom comparisons.
